crypto/x509/x_pubkey.c

```diff
@@ -248,7 +248,7 @@ int i2d_RSA_PUBKEY(RSA *a, unsigned char **pp)
         ASN1err(ASN1_F_I2D_RSA_PUBKEY, ERR_R_MALLOC_FAILURE);
         return -1;
     }
-    EVP_PKEY_set1_RSA(pktmp, a);
+    EVP_PKEY_set1_RSA(pktmp, (RSA *)a); /* up_ref() will be undone */
```

Comment:
Hmmmm....this cast is nasty (Same comment for other similar places)

Reply:
EVP_PKEY_set1_RSA actually modifies it, it increases the reference count.

Reply:
I consider 'nasty' a rather strong word in these cases since the type is not changed but 'just' the interpretation of the fields as being constant. 

Please note that the only modification that `EVP_PKEY_up_ref()`, `EVP_PKEY_set1_RSA()`, `EVP_PKEY_set1_DSA()`, and EVP_PKEY_set1_EC_KEY()` do to their second parameter is to increase its reference count on success, and exactly in this case the `up_ref()` is undone two lines later (indirectly via `EVP_PKEY_free()`), as I made explicit by the new comment(s). Therefore I thought the casts should be acceptable.

Nevertheless I've just found a solution that is more clean: not to use `EVP_PKEY_set1_RSA()` but  `EVP_PKEY_assign_RSA()`, which does not increase the reference count (and then of course to prevent `EVP_PKEY_free()` from decreasing it), and something analogous for the other public key types. Unfortunately the assignments to `pktmp->pkey.ptr` still make the compiler complain that `const` cannot be used here, but I think this is not actually the case here. What do you think?

Doing so I made also some stability improvements of `X509_PUBKEY_set()` and removed some weird type casts: `(char *)` that existed in `evp.h` ;-)

Reply:
IMO, casting shouldn't be used to simply hide the fact that an up reference is occurring. What happens when you pass a pointer to a variable that's resident in read-only memory (something the compiler will do if it can) and thus cause a segfault on attempts to modify data that is in the read only section?

Reply:
The RSA structure is opaque, i.e. you *must* create it dynamically using one of `RSA_new()` or `RSA_new_method()`.  I challenge you to find one that ends up in ro memory

Reply:
@mlindner, moreover, your comment is no more applicable to the current state of this PR - please note the yellow box above saying 'Outdated'.
Due to the solution I mentioned and pushed on Jan 21, the use of `EVP_PKEY_set1_RSA()` has been replaced by `EVP_PKEY_assign_RSA()` and the up and down of the reference count is gone.